### PR TITLE
Return 404 to UserAgents matching /bot/

### PR DESCRIPTION
**Pull request by jobwat**
snappass links on a chat (slack.com) were not working
Their preview request was consuming the password with the following User-Agent:
```
Slackbot-LinkExpanding 1.0 (+https://api.slack.com/robots)
```

### CODE REVIEW

File: snappass/main.py [hunk 1]
```diff
@@ -99,6 +100,8 @@ def handle_password():
 
 @app.route('/<password_key>', methods=['GET'])
 def show_password(password_key):
+    if re.search('bot', request.headers.get('User-Agent', '')):
```
Comment (nichochar) [inline snappass/main.py]: This is a good idea. Do you think we could move this out to a function called `request_is_valid(request)`, that returns a boolean? That way in the future we may want to append some logic in it.

**Comment by jobwat** [comment]
I did add a method as you asked, but not completely sure it gained readability!

Tell me if you want me to rebase this PR in 1 commit

And thanks for sharing your project!

**Comment by jameswthorne** [comment]
I too encountered this when I started using SnapPass. I ended up implementing this to get around it: #43 

Glad to see someone else thought of another method to get around this issue. Do you know if all the popular chat platforms that issue a GET the moment a URL is inputted into their text field will have a UserAgent with "bot" in it?

**Comment by jobwat** [comment]
@jameswthorne no I don't. The next unwanted User-Agent (or any criteria) will have to be added to the code.

But I believe /bot/ is a good start :)

**Comment by nichochar** [comment]
I prefer this solution to the Intermediary page @jameswthorne . But exactly as you guys said, I think we may append to the cases we ignore in the future, so I would prefer it if we could abstract that part of the code out in a function (see PR comments).

**Comment by jameswthorne** [comment]
Should the regex search for **bot** be replaced with a list containing all the UserAgent strings we want to 404?

I've collected the following UserAgent strings. I still need WhatsApp. I think iMessage impersonates facebook and twitter [Source](https://www.reddit.com/r/iOSProgramming/comments/4wcake/ios_10_imessage_user_agent_impersonates_facebook/).

- "facebookexternalhit/1.1" 
- "Slackbot-LinkExpanding 1.0 (+https://api.slack.com/robots)"
- "Twitterbot/1.0"

**Comment by nichochar** [comment]
Yes, matching a string present in the list sounds like a good design

**Comment by jameswthorne** [comment]
iMessage impersonates Facebook and Twitter. Very weird:

    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/601.2.4 (KHTML, like Gecko) Version/9.0.1 Safari/601.2.4 facebookexternalhit/1.1 Facebot Twitterbot/1.0"

I completely forgot about Skype, which appears to not do any sort of prefetching. Nor does Google Hangouts.

**Comment by jobwat** [comment]
So @jameswthorne, you're cool with my last commit?

**Comment by jobwat** [comment]
I completely understand your opinion

My view is slightly more lazy and optimistic: assuming that all the future newcomers will have a decently named User-Agent including the keyword `bot` and that no new browser will have it :)

We will never list all the annoying User-Agents, so I prefer covering a 'potential' amount of new User-Agent now with `bot` than listing all big sites names (where potentially no-one is using this service on)
And on my defence: try to find 1 existing browser User-Agent with `bot` ;)

@nichochar, you'll have the last word

**Comment by nichochar** [comment]
Sorry to keep on with the change requests, but I prefer the strict list:

- Slackbot
- facebookexternalhit
- Twitterbot
- Facebot
- WhatsApp

**Comment by jobwat** [comment]
I did implement the list as requested, and squashed commits to 1 clean one
:)

File: snappass/main.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import os
 import sys
 import uuid
+import re
```
Comment (jparise) [inline snappass/main.py]: nit: it would be nice to maintain the alphabetized order here

File: snappass/main.py [hunk 3]
```diff
@@ -78,6 +79,15 @@ def clean_input():
...
     return time_conversion[time_period], request.form['password']
 
+def request_is_valid(request):
+    """
+    Ensure the request validates the following:
+        - not made by some specific User-Agents (to avoid chat's preview feature issue)
+    """
+    known_sneaky_user_agents = ['Slackbot', 'facebookexternalhit', 'Twitterbot', 'Facebot', 'WhatsApp']
+    user_agents_regexp = "|".join(known_sneaky_user_agents)
+    return not re.search(user_agents_regexp, request.headers.get('User-Agent', ''))
```
Comment (jparise) [inline snappass/main.py]: Could we build and compile this regular expression object once at import-time?

```python
SNEAKY_USER_AGENTS = ('Slackbot', 'facebookexternalhit', 'Twitterbot', 'Facebot', 'WhatsApp')
SNEAKY_USER_AGENTS_RE = re.compile('|'.join(SNEAKY_USER_AGENTS))

def request_is_valid(request):
    return not SNEAKY_USER_AGENTS_RE.search(request.headers.get('User-Agent', ''))
```

File: tests.py [hunk 4]
```diff
@@ -73,6 +73,24 @@ def test_show_password(self):
         rv = self.app.get('/{0}'.format(key))
         self.assertTrue(password in rv.get_data(as_text=True))
 
+    def test_bots_denial(self):
+        """
+        Main known bots User-Agent should be denied access
+        """
+        password = "Bots can't access this"
+        key = snappass.set_password(password, 30)
+        a_few_sneaky_bots = [
```
Comment (jparise) [inline tests.py]: Should `facebot` also be included here? Perhaps this list should be build using variations of `SNEAKY_USER_AGENTS`.

Comment (jobwat) [inline tests.py]: I've add a FaceBot User-Agent to the list. I believe building the user-agent from our keyword list is not the best approach.

Thanks for the code fixes/propositions

File: snappass/main.py [hunk 5]
```diff
@@ -8,6 +9,8 @@
 from flask import abort, Flask, render_template, request
 
 
+SNEAKY_USER_AGENTS = ('Slackbot', 'facebookexternalhit', 'Twitterbot', 'Facebot', 'WhatsApp')
+SNEAKY_USER_AGENTS_RE = re.compile('|'.join(SNEAKY_USER_AGENTS))
```
Comment (jparise) [inline snappass/main.py]: Should we also specify `re.I` to request a case-insensitive match?

**Comment by nichochar** [comment]
Thanks for the PR, this is a good feature!

**Comment by jameswthorne** [comment]
Thanks for creating this @jobwat. This is a much cleaner method than the intermediary page I created.

I just tested it using Facebook Messenger, and it worked great.

**Comment by jobwat** [comment]
Glad you like it. Thank you guys for reviewing/editing/making it happen.

**Pull request merged**